### PR TITLE
feat(serverless-contracts): refactor output schemas

### PR DESCRIPTION
**Pull request by Sc0ra**

### CODE REVIEW

File: packages/serverless-contracts/src/contracts/apiGateway/features/fetchRequest.ts [hunk 1]
```diff
@@ -29,5 +29,8 @@ export const getFetchRequest = async <Contract extends ApiGatewayContract>(
     body: JSON.stringify(body),
   });
 
-  return response.json() as Promise<OutputType<Contract>>;
+  return {
+    statusCode: response.status,
+    body: (await response.json()) as OutputType<Contract>['body'],
+  } as OutputType<Contract>;
```
Comment (fargito) [inline packages/serverless-contracts/src/contracts/apiGateway/features/fetchRequest.ts:35]: 😍

File: packages/serverless-contracts/src/contracts/apiGateway/features/lambdaHandler.ts [hunk 2]
```diff
@@ -1,12 +1,17 @@
+/* eslint-disable complexity */
```
Comment (fargito) [inline packages/serverless-contracts/src/contracts/apiGateway/features/lambdaHandler.ts:1]: Why is this necessary?

File: packages/serverless-contracts/src/contracts/apiGateway/features/lambdaHandler.ts [hunk 3]
```diff
@@ -29,7 +34,19 @@ export const getApiGatewayHandler =
...
+        {
+          statusCode: number | string | symbol;
+          body: unknown;
+        }
+      >
+    > = OutputsType<Contract>,
+    Output extends {
+      statusCode: number | string | symbol;
+      body: unknown;
+    } = Exclude<O.UnionOf<Outputs>, undefined>,
```
Comment (fargito) [inline packages/serverless-contracts/src/contracts/apiGateway/features/lambdaHandler.ts:49]: Maybe this is too complex, now that we are accepting the breaking change

File: packages/serverless-contracts/src/contracts/apiGateway/features/lambdaHandler.ts (hunk 2 above)
Comment (Sc0ra) [reply packages/serverless-contracts/src/contracts/apiGateway/features/lambdaHandler.ts:1]: It was not, i removed it

**Review by fargito — APPROVED** [review]
Excellent work, thanks @Sc0ra and @StanHannebelle!

However I would prefer we create an alpha release in order to experiment this breaking change on our projects before the official release. I'll check this on Monday.

**Comment by fargito** [comment]
waiting for #417

**Review by fargito — CHANGES_REQUESTED** [review]
Careful about the commit titles, they should have an exclamation mark to set them as breaking

File: packages/serverless-contracts/src/types/http.ts [hunk 4]
```diff
@@ -1 +1,347 @@
+/* eslint-disable max-lines */
 export type HttpMethod = 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE';
+
+export enum StatusCodes {
```
Comment (adriencaccia) [inline packages/serverless-contracts/src/types/http.ts:4]: As the library does not only export tools related to HTTP, we should name it `HttpStatusCodes` to be really explicit.

**Review by adriencaccia — CHANGES_REQUESTED** [review]

**Comment by adriencaccia** [comment]
> Excellent work, thanks @Sc0ra and @StanHannebelle!
> 
> However I would prefer we create an alpha release in order to experiment this breaking change on our projects before the official release. I'll check this on Monday.

I took the liberty of changing the commit messages to explicit the breaking changes.

File: user-docs/documentation/docs/how-to-guides/3-migration-guides/4-output-schemas.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
+---
+sidebar_position: 4
+---
+
+# From outputSchema to outputSchemas
+
+## Why?
+
+Before version 0.25.0, in the `ApiGatewayContract`, the `outputSchema` key was used to specify the output schema of an Api Gateway integration.
```
Comment (adriencaccia) [inline user-docs/documentation/docs/how-to-guides/3-migration-guides/4-output-schemas.md:9]: Add a link to the release, even if it does not exist yet 🙈

File: user-docs/documentation/docs/how-to-guides/3-migration-guides/4-output-schemas.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
+---
+sidebar_position: 4
+---
+
+# From outputSchema to outputSchemas
```
Comment (adriencaccia) [inline user-docs/documentation/docs/how-to-guides/3-migration-guides/4-output-schemas.md:5]:
```suggestion
# From `outputSchema` to `outputSchemas`
```

File: user-docs/documentation/docs/how-to-guides/3-migration-guides/4-output-schemas.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+However, some users reported that they needed multiple output schemas for a single integration to handle different status codes, for error responses for example.
+Thus, we decided to add a `outputSchemas` to allow that and remove the `outputSchema` key, which introduced a breaking change.
+
+For more context, check out [the migration PR](https://github.com/swarmion/swarmion/pull/404).
+
+## How?
+
+Migrating from `outputSchema` to `outputSchemas` is quite straightforward:
+
+1. Upgrade serverless-contracts to version 0.25.0 or higher
```
Comment (adriencaccia) [inline user-docs/documentation/docs/how-to-guides/3-migration-guides/4-output-schemas.md:19]:
```suggestion
1. Upgrade `@swarmion/serverless-contracts` to version `0.25.0` or higher
```

**Review by adriencaccia — CHANGES_REQUESTED** [review]

File: packages/serverless-contracts/src/contracts/apiGateway/features/openApiDocumentation.ts [hunk 8]
```diff
@@ -28,7 +29,9 @@ export const getOpenApiDocumentation = <
           // This cast is done because there is differences between JsonSchema and OpenAPIV3.SchemaObject specs
           // It may be fixed later
           // @ref https://swagger.io/specification/
-          schema: contract.outputSchema as OpenAPIV3.SchemaObject,
+          schema: contract.outputSchemas[
+            HttpStatusCodes.OK
+          ] as OpenAPIV3.SchemaObject,
```
Comment (fargito) [inline packages/serverless-contracts/src/contracts/apiGateway/features/openApiDocumentation.ts:34]: Add a comment here for improvement: given this PR, we will be able to be much more precise on the openAPI schema 🚀

File: user-docs/documentation/docs/getting-started/5-add-a-new-serverless-contract-library.mdx [hunk 9]
```diff
@@ -110,7 +110,10 @@ in order to have live type updates without the need to package the contracts lib
 Now lets write our first contract to define a get user operation.
 
 ```typescript title="contracts/users-contracts/src/contracts/getUser.ts"
-import { ApiGatewayContract } from '@swarmion/serverless-contracts';
+import {
+  ApiGatewayContract,
+  HttpStatusCodes,
+} from '@swarmion/serverless-contracts';
```
Comment (fargito) [inline user-docs/documentation/docs/getting-started/5-add-a-new-serverless-contract-library.mdx:116]: We need to update the generators as well

**Review by fargito — APPROVED** [review]

**Review by adriencaccia — APPROVED** [review]

**Pull request merged**